### ISSUE TITLE
Passing class or id of script tag to tags generated by Mathjax

### ISSUE THREAD

**Issue by luc-j-bourhis** [mathjax/MathJax#2041]
Mathjax can be styled with CSS using classes "Mathjax_Display" and "Mathjax" but they are all or nothing. Especially if the HTML featuring Mathjax script tags is generated from a different format, e.g. an extended Markdown such as kramdown, then there is no way to pass a class or an id from that Markdown source to the final rendered equation, as a class or an id set on the Mathjax script tag is not passed onto the div's and span's generated by Mathjax. It would be nice if Mathjax was passing those.

Note that the HTML standard allows class and id on any tag (they are among the so-called global attributes), and that includes <script>.

**Comment by pkra** [comment]
Oh, nevermind, you want something on the script tag. Sorry for the noise.

**Comment by dpvc** [comment]
> there is no way to pass a class or an id from that Markdown source to the final rendered equation
> 

If you are writing your math using LaTeX notation, you can use the \class or \cssId macros to add classes and ids to portions of your mathematics.  For example,

	\(x \class{operator}{+} y\)

would produce output that has class="operator" applied to the HTML that represents the plus sign. So you could use 

	.operator {color: red}

to style the operator(s) red.  Of course, you can enclose an entire expression with a class or id:

	\(\cssId{eq1}{\sqrt{\frac{x+1}{x-1}}}\)

(though the element with the id="eq1" will be partway down into the math tree, since there are some MathML elements surrounding the square root, and the id is on the msqrt element in the MathML tree that the TeX has been converted into).

> a class or an id set on the Mathjax script tag is not passed onto the div's and span's generated by Mathjax
> 

That is not actually true.  MathJax actually sets ids on all the script tags that it uses internally (if they don't already have one) and uses those ids to identify each tag and tie that to its associated output.  Note, however, that ids must be unique within a document, and so MathJax can't reuse an id from the script directly on the output, as that would produce duplicate ids.  Instead, MathJax appends "-Frame" onto the id from the script element and uses that as the id of the span or div that contains the mathematics.

So if you have

	<script type="math/tex" id="eq1">x+1</script>

then MathJax will produce something like

	<span id="eq1-Frame" class="mjx-chtml MathJax_CHTML" tabindex="0" style="font-size: 119%;">...</span>

The actual result depends on the output jax that you are using, but they all produce something of this kind.  So you could use "#eq1-Frame" as the selector to style the contents of the math from the script with id="eq1".

If you are planning to do styling that includes background colors or borders and such, then you will probably want to use the CommonHTML output jax rather than the older HTML-CSS output jax, as the former uses correct bounding boxes for every component of the expression, while the latter uses spans whose background may not fully enclose their contents (which is how normal spans work when their contents are larger than the line height).  For HTML-CSS, you might have to use something like

	#eq1-Frame .math {
	    background-color: red;
	    padding: .3em;
	}

to add some padding and background color to HTML-CSS output.

Davide

**Comment by dpvc** [comment]
> there is no way to pass a class or an id from that Markdown source to the final rendered equation

If you are writing your math using LaTeX notation, you can use the `\class` or `\cssId` macros to add classes and ids to portions of your mathematics.  For example,

    \(x \class{operator}{+} y\)

would produce output that has `class="operator"` applied to the HTML that represents the plus sign. So you could use 

    .operator {color: red}

to style the operator(s) red.  Of course, you can enclose an entire expression with a class or id:

    \(\cssId{eq1}{\sqrt{\frac{x+1}{x-1}}}\)

(though the element with the `id="eq1"` will be partway down into the math tree, since there are some MathML elements surrounding the square root, and the id is on the `msqrt` element in the MathML tree that the TeX has been converted into).

> a class or an id set on the Mathjax script tag is not passed onto the div's and span's generated by Mathjax

That is not actually true.  MathJax actually sets ids on all the script tags that it uses internally (if they don't already have one) and uses those ids to identify each tag and tie that to its associated output.  Note, however, that ids must be unique within a document, and so MathJax can't reuse an id from the script directly on the output, as that would produce duplicate ids.  Instead, MathJax appends `-Frame` onto the id from the script element and uses that as the id of the span or div that contains the mathematics.

So if you have

    <script type="math/tex" id="eq1">x+1</script>

then MathJax will produce something like

    <span id="eq1-Frame" class="mjx-chtml MathJax_CHTML" tabindex="0" style="font-size: 119%;">...</span>

The actual result depends on the output jax that you are using, but they all produce something of this kind.  So you could use `#eq1-Frame` as the selector to style the contents of the math from the script with `id="eq1"`.

If you are planning to do styling that includes background colors or borders and such, then you will probably want to use the CommonHTML output jax rather than the older HTML-CSS output jax, as the former uses correct bounding boxes for every component of the expression, while the latter uses spans whose background may not fully enclose their contents (which is how normal spans work when their contents are larger than the line height).  For HTML-CSS, you might have to use something like

    #eq1-Frame .math {
        background-color: red;
        padding: .3em;
    }

to add some padding and background color to HTML-CSS output.

**Comment by luc-j-bourhis** [comment]
I realise I never thanked you for your explanations! Very helpful!

**Comment by dpvc** [comment]
No problem.  Glad it was able to help.
